### PR TITLE
CA App context

### PR DESCRIPTION
**Pull request by norisalsaadie**
Migrating the app context component and tests.

components/ca-app-context.js -> migrate from old repo to ES6.
test/ca-app-context-test.html -> migrate tests from old repo.

### CODE REVIEW

File: test/ca-app-context-test.html [hunk 1]
```diff
@@ -0,0 +1,97 @@
+<!doctype html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <title>CA DIALOG | Web Lab Common UI</title>
```
Comment (kevmarchant) [inline test/ca-app-context-test.html]: title needs changing...

Comment (norisalsaadie) [inline test/ca-app-context-test.html]: Oops

Comment (norisalsaadie) [inline test/ca-app-context-test.html]: done

**Comment by ElementForge** [comment]
OK, I've looked at the code but I don't really feel I understand the context in which ca-app-context is going to be used. I'm sure someone could explain this quickly. But if that's the case, then that explanation should be placed in the code to help others.

Also, I'm not sure where the event string 'ca-config-component' comes into this. Who defines all these custom strings? How many event names are there going to be? Are they dreamt up or is there some purpose to the event names. Would a better name be 'ca-app-context-config' or am I missing something here...?

**Comment by norisalsaadie** [comment]
@ElementForge https://bitbucket.org/carepo/tc-web-ui/src/672f5b14d18def2f12494fb4eb72b39673f4e4c4/app/js/app.js?at=master&fileviewer=file-view-default#app.js-35 is the current use case. 

We need to enable a remote-image-capture but only when PI is using the application, but an ordinary image-capture when using a regular desktop/laptop. To avoid hard coding things since most schemas will have a level of nesting, e.g. ca-crud -> ca-form -> ca-image-capture. This allows components to fire an event (doesn't matter who consumes it) and gives the application the chance to configure it e.g. give it some specific values. 

It's essential a lower level version of what angular 1.x is doing with event broadcasting. Ideally, this will evolve to be part of ca-app or something similar.

The current flow is this:

<ca-app-context> lives in the root of an app and listens to ca-config-component.

Any component can fire an event, and be configured. (see https://bitbucket.org/carepo/weblab-common-ui/src/dd25c7761dbb8af3664d0b96be08ee1153ac022e/components/ca-remote-image-capture.js?at=master&fileviewer=file-view-default#ca-remote-image-capture.js-74).

I hope this makes sense. If not then feel free to ask.

In any case, the objective of stage 1 is to migrate everything over to amd modules, add TODOs for missing tests, and convert to ES6. There is to be another step which includes adding missing tests, and removing/merging components

File: test/ca-app-context-test.html [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                  },
+                  'bubbles': true
+                });
+
+                // act
+                appContext.setAttribute('data-rpi', true);
+                testElement.dispatchEvent(event);
+
+                // verify
+                expect(appContext.getAttribute('data-rpi')).to.be.defined;
```
Comment (tapina) [inline test/ca-app-context-test.html]: I agree that this is a direct port, but it needs a TODO as it's not actually testing anything useful. It ought to be checking the attribute on the testElement instead.

File: test/ca-app-context-test.html [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                      component: testElement
+                  },
+                  'bubbles': true
+                });
+
+                // act
+                testElement.dispatchEvent(event);
+
+                // event will be consumed on the next event loop
+                setTimeout(() => expect(testElement.getAttribute('data-rpi')).to.equal('yes~yes'), 0);
```
Comment (tapina) [inline test/ca-app-context-test.html]: Ah, so this is an extra test which actually tests something then?

Comment (norisalsaadie) [inline test/ca-app-context-test.html]: Yep :)

**Pull request merged**